// @langchain/core/dist/testing/fake_model_builder.cjs
const require_base = require('../messages/base.cjs');
const require_ai = require('../messages/ai.cjs');
const require_base$1 = require('../runnables/base.cjs');
require('../messages/index.cjs');
const require_language_models_chat_models = require('../language_models/chat_models.cjs');

//#region src/testing/fake_model_builder.ts
function deriveContent(messages) {
	return messages.map((m) => m.text).filter(Boolean).join("-");
}
let idCounter = 0;
function nextToolCallId() {
	idCounter += 1;
	return `fake_tc_${idCounter}`;
}
/**
* A fake chat model for testing, created via {@link fakeModel}.
*
* Queue responses with `.respond()` and `.respondWithTools()`, then
* pass the instance directly wherever a chat model is expected.
*
* Each builder method queues a model response consumed in order per `invoke()`:
*
* `.respond(entry)` — enqueue a `BaseMessage`, `Error`, or factory function.
*
* `.respondWithTools(toolCalls[])` — enqueue an `AIMessage` with the given
* tool calls. Content is derived from the input messages automatically.
*
* Both can be mixed freely in one chain. When all queued responses are
* consumed, further invocations throw.
*
* Additional configuration:
* - `.alwaysThrow(error)` — every call throws (overrides the queue)
* - `.structuredResponse(value)` — value returned by `withStructuredOutput()`
*
* The model records all invocations in `.calls` / `.callCount`.
*/
var FakeBuiltModel = class FakeBuiltModel extends require_language_models_chat_models.BaseChatModel {
	queue = [];
	_alwaysThrowError;
	_structuredResponseValue;
	_tools = [];
	_callIndex = 0;
	_calls = [];
	get calls() {
		return this._calls;
	}
	get callCount() {
		return this._calls.length;
	}
	constructor() {
		super({});
	}
	_llmType() {
		return "fake-model-builder";
	}
	_combineLLMOutput() {
		return [];
	}
	respond(entry) {
		if (typeof entry === "function") this.queue.push({
			kind: "factory",
			factory: entry
		});
		else if (require_base.BaseMessage.isInstance(entry)) this.queue.push({
			kind: "message",
			message: entry
		});
		else this.queue.push({
			kind: "error",
			error: entry
		});
		return this;
	}
	respondWithTools(toolCalls) {
		this.queue.push({
			kind: "toolCalls",
			toolCalls: toolCalls.map((tc) => ({
				name: tc.name,
				args: tc.args,
				id: tc.id ?? nextToolCallId(),
				type: "tool_call"
			}))
		});
		return this;
	}
	alwaysThrow(error) {
		this._alwaysThrowError = error;
		return this;
	}
	structuredResponse(value) {
		this._structuredResponseValue = value;
		return this;
	}
	bindTools(tools) {
		const merged = [...this._tools, ...tools];
		const next = new FakeBuiltModel();
		next.queue = this.queue;
		next._alwaysThrowError = this._alwaysThrowError;
		next._structuredResponseValue = this._structuredResponseValue;
		next._tools = merged;
		next._calls = this._calls;
		next._callIndex = this._callIndex;
		return next.withConfig({});
	}
	withStructuredOutput(_params, _config) {
		const { _structuredResponseValue } = this;
		return require_base$1.RunnableLambda.from(async () => {
			return _structuredResponseValue;
		});
	}
	async _generate(messages, options, _runManager) {
		this._calls.push({
			messages: [...messages],
			options
		});
		const currentCallIndex = this._callIndex;
		this._callIndex += 1;
		if (this._alwaysThrowError) throw this._alwaysThrowError;
		const entry = this.queue[currentCallIndex];
		if (!entry) throw new Error(`FakeModel: no response queued for invocation ${currentCallIndex} (${this.queue.length} total queued).`);
		if (entry.kind === "error") throw entry.error;
		if (entry.kind === "factory") {
			const result = entry.factory(messages);
			if (!require_base.BaseMessage.isInstance(result)) throw result;
			return { generations: [{
				text: "",
				message: result
			}] };
		}
		if (entry.kind === "message") return { generations: [{
			text: "",
			message: entry.message
		}] };
		const content = deriveContent(messages);
		return {
			generations: [{
				text: content,
				message: new require_ai.AIMessage({
					content,
					id: currentCallIndex.toString(),
					tool_calls: entry.toolCalls.length > 0 ? entry.toolCalls.map((tc) => ({
						...tc,
						type: "tool_call"
					})) : void 0
				})
			}],
			llmOutput: {}
		};
	}
};
/**
* Creates a fake chat model for testing.
*
* @example
* ```typescript
* const model = fakeModel()
*   .respondWithTools([{ name: "search", args: { query: "weather" } }])
*   .respond(new AIMessage("Sunny and warm."));
*
* const r1 = await model.invoke([new HumanMessage("What's the weather?")]);
* // r1.tool_calls[0].name === "search"
*
* const r2 = await model.invoke([new HumanMessage("Thanks")]);
* // r2.content === "Sunny and warm."
* ```
*/
function fakeModel() {
	return new FakeBuiltModel();
}

//#endregion
exports.fakeModel = fakeModel;
